README
NAME
    Sim::Dispatcher - Event dispatcher for Sim

VERSION
    This document describes Sim::Dispatcher 0.02 released on June 2, 2007.

SYNOPSIS
        use Sim::Dispatcher;
        use Sim::Clock;

        my $clock = Sim::Clock->new;
        # you can also use your own Clock instance here
        my $engine = Sim::Dispatcher->new(clock => $clock);

        # Example 1: Static scheduling

        $engine->schedule(
           0 => sub { print $engine->now, ": morning!\n" },
           1 => sub { print $engine->now, ": afternoon!\n" },
           5 => sub { print $engine->now, ": night!\n" },
        );
        $engine->run( duration => 50 );
        # or Sim::Dispatcher->run( fires => 5 );

        $engine->reset();

        # Example 2: Dynamic (recursive) scheduling

        my ($count, $handler);

        # event handler:
        $handler = sub {
            $count++;
            my $time_for_next = $engine->now() + 2;
            $engine->schedule(
                $time_for_next => $handler,
            );
        };
        # only schedule the "seed" event
        $engine->schedule(
            0.5 => $handler,
        );
        $engine->run( fires => 5 );
        print "count: $count\n";  # 5
        print "now: ", $engine->now(), "\n";  # 8

DESCRIPTION
    This class implements the most important component in the whole Sim
    library, the event dispatcher. Basically, every activites should be
    coordinated by this dispatcher. Every other objects in a simulator
    either register an event scheduled to happen at some point in the
    "future", or iterate through the dispatching steps.

METHODS
    "$obj = Sim::Dispatcher->new( clock => $clock)"
        Object constructor accepting one mandatory named argument $clock
        which is an instance of classes like Sim::Clock.

    "$obj->schedule( $time => $handle, ... )"
        You can use this method to register events scheduled for the future,
        where $time is the timestamp and $handle is an anonymous sub which
        will be invoked by the dispatcher when the simulation time is at
        $time.

    "$obj->run( duration => $time, fires => $count )"
        Runs the dispatcher according to the time duration and event firing
        count. both of these named parameters are optional. When none is
        specified, "fires => 100_000_000" will be assumed.

    "$obj->fire_next()"
        This method allows you to iterate through the dispatcher running
        process yourself. You should only call "fire_next" by hand if you've
        found the limitation criteria given by the "run" method can't fit
        your needs.

    "$obj->now()"
        Reads the value of the simulation time.

    "$obj->time_of_next()"
        Gets the timestamp of the next (or nearest) coming event, which is
        always a bit greater or equal to "now".

    "$obj->reset()"
        Clears the internal event queue of the dispatcher and resets the
        internal simulation clock too.

CONCURRENCY ISSUES
    If two events have exactly the same timestamp, say, 1.5, then the one
    registered earlier will be fired first.

AUTHOR
    Agent Zhang <agentzh@gmail.com>

COPYRIGHT
    Copyright 2006, 2007 by Agent Zhang. All rights reserved.

    This library is free software; you can modify and/or modify it under the
    same terms as Perl itself.

SEE ALSO
    Sim::Clock, Sim.



Changes
Changes for Sim

0.02 - 2007-06-02
* fixed the no_index call in Makefile.PL
* added $VERSION var to Sim::Clock

0.01 - 2007-06-02
* initial CPAN release



lib/Sim.pm
package Sim;

use strict;
use warnings;
use vars qw( $AUTOLOAD );

use Sim::Dispatcher;
use Sim::Clock;

our $VERSION = '0.02';

our ($Clock, $Dispatcher);

sub AUTOLOAD {
    my $self = shift;
    my $method = $AUTOLOAD;

    $method =~ s/.*:://;
    return if $method eq 'DESTROY';

    $Clock ||= Sim::Clock->new;
    $Dispatcher ||= Sim::Dispatcher->new(clock => $Clock);
    $Dispatcher->$method(@_);
}

1;
__END__

=head1 NAME

Sim - Simulator engine for discrete events

=head1 VERSION

This document describes Sim 0.02 released on 2 June, 2007.

=head1 SYNOPSIS

    use Sim;

    Sim->schedule(
        0.2 => sub { print "Hi\n" },
        0.4 => sub {
            Sim->schedule(
                0.5 => sub { print "Wow!\n" },
                Sim->now + 0.2 => sub { print "Hello!\n" },
            );
        },
        0.5 => sub { print "now is ", Sim->now, "\n"; },
    );

    Sim->run( duration => 1.0 );  # upper-limit for simulation time

    # OR: Sim->run( fires => 15 );  # upper-limit for number of event fires

    print "now is ", Sim->now, "\n";  # now is 0.6

=head1 DESCRIPTION

Sim is a general-purpose discrete event simulator engine written in pure
Perl. It was originally developed as the run-core of a sequential/conbinational
logic circuit simulator named Tesla.

The Sim class is just a static class wrapping around a L<Sim::Dispatcher>
instance and a L<Sim::Clock> instance. I used AUTOLOAD to do the magic instead
of writing a lot of boring code. It's expected that using L<Sim::Dispatcher>
directly can be a bit faster but is certainly less convenient.

If you want to use a different clock model with vectorized time read,
say a [sec, delta], which is found in a lot of EDA simulators, then you
should use L<Sim::Dispatcher> instead of L<Sim> and define your clock class
yourself.

See L<Sim::Dispatcher> for more information.

=head1 STATIC METHODS

All the methods of L<Sim::Dispatcher> are available for this class, but
they are exposed as static methods only :)

=head1 VERSION CONTROL

You can always get the latest version of the source from the following
Subversion repository:

L<http://svn.openfoundry.org/sim/>

which has anonymous access to all.

If you do want a commit bit and become a coauthor, please let me know :)

=head1 CODE COVERAGE

I use Devel::Cover to test the code coverage of the test suite:

---------------------------- ------ ------ ------ ------ ------ ------ ------
File                           stmt   bran   cond    sub    pod   time  total
---------------------------- ------ ------ ------ ------ ------ ------ ------
blib/lib/Sim.pm               100.0   50.0   66.7  100.0    n/a    5.7   91.7
blib/lib/Sim/Clock.pm         100.0   83.3    n/a  100.0  100.0    9.0   97.3
blib/lib/Sim/Dispatcher.pm     94.6   75.0  100.0  100.0  100.0   31.1   92.2
---------------------------- ------ ------ ------ ------ ------ ------ ------

=head1 AUTHOR

Agent Zhang E<lt>agentzh@gmail.comE<gt>

=head1 COPYRIGHT

Copyright 2006, 2007 by Agent Zhang. All rights reserved.

This library is free software; you can modify and/or modify it under the
same terms as Perl itself.

=head1 SEE ALSO

L<Sim::Dispatcher>, L<Sim::Clock>.



lib/Sim/Clock.pm
package Sim::Clock;

use Carp qw(croak carp);
use strict;
use warnings;

our $VERSION = '0.02';

sub new ($$) {
	my $class = ref $_[0] ? ref shift : shift;
	my $now = @_ ? shift : 0;
	bless {
		now => $now,
	}, $class;
}

sub now ($) {
	$_[0]->{now};
}

sub push_to ($$) {
    my ($self, $time) = @_;
	if ($time < $self->now) {
        carp "error: Can't push your time back, sir";
        return 0;
    }
    $self->{now} = $time;
    return 1;
}

sub reset ($) {
    $_[0]->{now} = 0;
}

1;
__END__

=head1 NAME

Sim::Clock - Simulation clock used by the dispatcher

=head1 VERSION

This document describes Sim::Clock 0.02 released on
2 June, 2007.

=head1 SYNOPSIS

    use Sim::Clock;
    my $clock = Sim::Clock->new(0);
    $clock->push_to(5.6);
    print $clock->now;  # 5.6
    $clock->push_to(3); # exception!
    $clock->reset();

=head1 DESCRIPTION

This class offers a simulation clock for L<Sim::Dispatcher>. Basically
you needn't create your own clock at all since L<Sim::Dispatcher> always
creates one internally and you seldom or never need more than one clock
in your simulator. But you really feel the need to do something fancy,
simulating quantum mechanics for example, here is the rope.

=head1 METHODS

=over

=item C<< $obj->new( $init_time ? ) >>

Ths is the constructor for C<Sim::Clock> objects. The C<$init_time> argument
specifies the initial time read of the clock and can be omitted.

=item C<< $obj->now() >>

Reads the current time from the clock.

=item C<< $obj->push_to($time) >>

Push the clock to the specified timestamp C<$time>, which can't be earlier than
the value obtained by the C<now> method.

=item C<< $obj->reset() >>

Resets the clock to time 0.

=back

=head1 AUTHOR

Agent Zhang E<lt>agentzh@gmail.comE<gt>

=head1 COPYRIGHT

Copyright 2006 by Agent Zhang. All rights reserved.

This library is free software; you can modify and/or
modify it under the same terms as Perl itself.

=head1 SEE ALSO

L<Sim::Dispatcher>, L<Sim>.



lib/Sim/Dispatcher.pm
package Sim::Dispatcher;

use strict;
use warnings;

our $VERSION = '0.02';

use Carp qw( carp croak );

our $DEBUG = 0;

sub new {
    my $self = ref $_[0] ? ref shift : shift;
    my %opts = @_;
    croak "No clock given" if !$opts{clock};
    bless {
        clock => $opts{clock},
        queue => [],
    }, $self;
}

sub now {
    $_[0]->{clock}->now;
}

sub schedule {
    my $self = shift;
    my %events = @_;
    while (my ($time, $handle) = each %events) {
        if ($time < $self->now) {
            carp "out-dated event [$time => $handle] ignored";
            next;
        }
        $self->_insert_event([$time => $handle]);
    }
}

sub _insert_event {
    my ($self, $event) = @_;
    my $queue = $self->{queue};
    for (my $i = 0; $i < @$queue; $i++) {
        if ($event->[0] < $queue->[$i]->[0]) {
            splice( @$queue, $i, 0, $event );
            return;
        }
    }
    push @$queue, $event;
}

sub fire_next ($) {
    my $self = shift;
    my $queue = $self->{queue};
    my $clock = $self->{clock};
    return undef if @$queue == 0;
    my $event = shift @$queue;
    my ($time, $handle) = @$event;
    my $now = $self->now;
    if ($time >= $now) {
        $clock->push_to($time);
        $handle->();
    } else {
        die "Clock modified outside of the dispatcher: next event is at $time while now is $now";
    }
    return 1;
}

sub run ($@) {
    my $self = shift;
    my %opts = @_;
    my $end_time = $self->now + $opts{duration} if defined $opts{duration};
    my $fires    = $opts{fires} || 100_000_000;
    my $i = 0;
    while (1) {
        #warn "run: next!";
        last if ++$i > $fires;
        my $t = $self->time_of_next;
        last if !defined $t or (defined $end_time and $t > $end_time);
        $self->fire_next;
    }
}

sub time_of_next ($) {
    my $self = shift;
    my $queue = $self->{queue};
    return @$queue ? $queue->[0]->[0] : undef;
}

sub reset ($) {
    my $self = shift;
    $self->{queue} = [];
    $self->{clock}->reset();
}

1;
__END__

=head1 NAME

Sim::Dispatcher - Event dispatcher for Sim

=head1 VERSION

This document describes Sim::Dispatcher 0.02 released
on June 2, 2007.

=head1 SYNOPSIS

    use Sim::Dispatcher;
    use Sim::Clock;

    my $clock = Sim::Clock->new;
    # you can also use your own Clock instance here
    my $engine = Sim::Dispatcher->new(clock => $clock);

    # Example 1: Static scheduling

    $engine->schedule(
       0 => sub { print $engine->now, ": morning!\n" },
       1 => sub { print $engine->now, ": afternoon!\n" },
       5 => sub { print $engine->now, ": night!\n" },
    );
    $engine->run( duration => 50 );
    # or Sim::Dispatcher->run( fires => 5 );

    $engine->reset();

    # Example 2: Dynamic (recursive) scheduling

    my ($count, $handler);

    # event handler:
    $handler = sub {
        $count++;
        my $time_for_next = $engine->now() + 2;
        $engine->schedule(
            $time_for_next => $handler,
        );
    };
    # only schedule the "seed" event
    $engine->schedule(
        0.5 => $handler,
    );
    $engine->run( fires => 5 );
    print "count: $count\n";  # 5
    print "now: ", $engine->now(), "\n";  # 8

=head1 DESCRIPTION

This class implements the most important component in the
whole Sim library, the event dispatcher. Basically, every
activites should be coordinated by this dispatcher.
Every other objects in a simulator either register an
event scheduled to happen at some point in the "future",
or iterate through the dispatching steps.

=head1 METHODS

=over

=item C<< $obj = Sim::Dispatcher->new( clock => $clock) >>

Object constructor accepting one mandatory named
argument C<$clock> which is an instance of classes like
L<Sim::Clock>.

=item C<< $obj->schedule( $time => $handle, ... ) >>

You can use this method to register events scheduled for the future, where
$time is the timestamp and $handle is an anonymous sub which will be invoked
by the dispatcher when the simulation time is at $time.

=item C<< $obj->run( duration => $time, fires => $count ) >>

Runs the dispatcher according to the time duration and event firing count.
both of these named parameters are optional. When none is specified,
C<< fires => 100_000_000 >> will be assumed.

=item C<< $obj->fire_next() >>

This method allows you to iterate through the dispatcher running process yourself.
You should only call C<fire_next> by hand if you've found the limitation criteria
given by the C<run> method can't fit your needs.

=item C<< $obj->now() >>

Reads the value of the simulation time.

=item C<< $obj->time_of_next() >>

Gets the timestamp of the next (or nearest) coming event, which is always a bit
greater or equal to "now".

=item C<< $obj->reset() >>

Clears the internal event queue of the dispatcher and resets the internal simulation
clock too.

=back

=head1 CONCURRENCY ISSUES

If two events have exactly the same timestamp, say, 1.5,
then the one registered earlier will be fired first.

=head1 AUTHOR

Agent Zhang E<lt>agentzh@gmail.comE<gt>

=head1 COPYRIGHT

Copyright 2006, 2007 by Agent Zhang. All rights reserved.

This library is free software; you can modify and/or modify it under the same terms as
Perl itself.

=head1 SEE ALSO

L<Sim::Clock>, L<Sim>.

